docs/proposals/multiple-sources-for-applications.md

```diff
@@ -0,0 +1,225 @@
+
+### Non-goals
+* Allow reconciliation from Argo CD Application resources that are located at various sources. We believe this would be possible with some adaptions in the reconcilation workflow.
+
+## Proposal
+
+### Add new `sources` field in Application Spec
+
+The proposal is to add a new field `sources` which would allow users to input list of `ApplicationSource`s. We would mark field `source` as deprecated and would override the details under `source` with details under `sources` field.
+
+```yaml
+spec:
+  source:
+    repoURL: https://charts.bitnami.com/bitnami
+    targetRevision: 8.5.8
+    helm:
+    valueFiles:
+        - values.yaml
+    chart: mysql
+  sources:                                          # new field
+    - repoURL: https://charts.bitnami.com/bitnami
+      targetRevision: 8.5.8
+      helm:
+        valueFiles:
+          - values.yaml
+      chart: mysql
+```
+
+### Add `externalValues` field in helm section of Application Spec
+
+Along with new `sources` field, add a new field for accepting external value files `externalValueFiles` under the `helm` section of each ApplicationSource.
+
+```yaml
+sources:
+    - repoURL: https://charts.bitnami.com/bitnami
+      targetRevision: 8.5.8
+      helm:
+        valueFiles:
+          - values.yaml
+        externalValueFiles:                         # new field
+          - repoURL: https://github.com/KaiReichart/argo-test-values.git
+            targetRevision: main
+            valueFiles:
+              - values.yaml
+      chart: mysql
+```
+
+### Combined Example Application yaml
+
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: grafana
+  namespace: argocd
+spec:
+  destination:
+    namespace: monitoring
+    server: https://some.k8s.url.com:6443
+  project: default
+  source:
+    repoURL: https://charts.bitnami.com/bitnami
+    targetRevision: 8.5.8
+    helm:
+    valueFiles:
+        - values.yaml
+    chart: mysql
+  sources:                                          # new field
+    - repoURL: https://charts.bitnami.com/bitnami
+      targetRevision: 8.5.8
+      helm:
+        valueFiles:
+          - values.yaml
+        externalValueFiles:                         # new field
+          - repoURL: https://github.com/KaiReichart/argo-test-values.git
+            targetRevision: main
+            valueFiles:
+              - values.yaml
+      chart: mysql
+    - repoURL: https://charts.bitnami.com/bitnami1
+      targetRevision: 8.5.8
+      helm:
+        valueFiles:
+          - values.yaml
+        externalValueFiles:
+          - repoURL: https://github.com/KaiReichart/argo-test-values.git
+            targetRevision: main
+            valueFiles:
+              - values.yaml
+      chart: mysql
```

Comment:
What's the expected behavior if two sources produce the same resource (i.e. two resources of the same kind with the same name)?

I think the same problem exists for single-source Applications. My guess is the last resource in the generated manifest "wins". If that's the case, I feel like the last resource from the last `source` (as determined by order in the Application manifest) should win.

Reply:
Most of the Helm-based applications will have the top-level values-files included into the chart and one or more low-level values-files provided via Git with the low-level values files overriding exactly the same settings as specified in the top-level template. 

This is why it is not possible to position them on the same level (unordered list), without specifying the order / priority. 

I see two options to resolve this:
 1. (indirect) Settings from all git-based values-files are processed after settings from the Helmchart-based values files which will cover 90% of the use cases.
 2. (direct) Add another section, in addition to **source** that will provide the repository (or a list of repositories) with the low-level (overriding) values-files. You can name it **override** or **detail**.

Some background: A typical Helm-based application is based on the external Helm-chart that already includes a default values file. At the deployment step, the application-specific values-file provided from Git overrides these settings. Some companies also have so-called stage-specific values files also provided from Git and the processing order looks like this:
External Helm Chart (+values) -> Stage-specific values-file from Git -> Application-specific values-file from Git.

Reply:
As far as I see it - this is solved by the way it already works. The order in which values files is listed in application.yaml - is the order in which its given to helm command (--values valuefile1 --values valuefile2 ...) - and this way Helms normal handling of values file ordering takes care of this. Which means the LAST values file overrides the settings from the previous ones. That is how it works now - I see no reason this should change anything. The values in the subchart (default values) is at the top of the list - hence any value given in the specified values files - overrides them.

Reply:
If the order is enforced and follows the order in the Application configuration, it should solve the concern above.
Thank you for clarifying!